### DISCUSSION TITLE
Building Archive didn't work successfully

### DISCUSSION THREAD

**Discussion by devahmedshendy** [swiftyfinch/Rugby#138] [General]
Hi @swiftyfinch ,
I am opening this discussion hoping this can be an issue that I can solve.

I was working fine with Rugby in development, then after I was ready to build the archive, it failed with following error:

![Screen Shot 2022-02-28 at 7 20 04 AM](https://user-images.githubusercontent.com/22206851/155928641-c25421c9-8e28-4ec8-93b8-9f73699e68bc.png)

Also I checked the warnings for this extension and found it complains about all pods from my project, like the following:
```
ld: warning: directory not found for option '-F/Project/Pods/../.rugby/build/Release-iphoneos/DifferenceKit'
```

**Comment by devahmedshendy** [comment]
## Did you already try Rugby doctor?
Rugby works fine for development, but I did `rugby doctor` and also upgraded to 1.17.0, and the issue still existed.

## Describe the bug
When I try to archive the project, it fails saying `Framework not found Alamofire`

## To Reproduce
I don't know how to create extension like `OneSignalNotificationServiceExtension` in my project.

Steps to reproduce the behavior:
1. Install Pods
`
pod deintegrate
pod install
`
2. Run Rugby
```
rugby
```
3. From Xcode click `Product` > `Archive`
4. It should failed saying `Framework not found Alamofire`

## Expected behavior
To archive the project that's already managed by rugby without any issue successfully.

## Screenshots
![Screen Shot 2022-02-28 at 7 20 04 AM](https://user-images.githubusercontent.com/22206851/155931489-cda4ea5d-1a2a-4b0d-a778-e09a2af8a658.png)

## Environment (please complete the following information):
 - Version 1.17.0
 - Logs (use `rugby doctor`)
 ```
 cat .rugby/rugby.log 
* Rugby version: 1.17.0
* Command: rugby 
* Xcode 13.2.1 / Build version 13C100
* Apple Swift version 5.5.2 (swiftlang-1300.0.47.5 clang-1300.0.29.30)
* CPU: Intel(R) Core(TM) i5-3210M CPU @ 2.50GHz
* Project: Waffer
* Git branch: main
---------------------------------
Prepare Read project ✓
Prepare Pods (28):
Prepare * Alamofire
Prepare * DifferenceKit
Prepare * ECSlidingViewController
Prepare * Firebase
Prepare * FirebaseAnalytics
Prepare * FirebaseAuth
Prepare * FirebaseCore
Prepare * FirebaseCoreDiagnostics
Prepare * FirebaseDatabase
Prepare * FirebaseInstallations
Prepare * Google-Maps-iOS-Utils
Prepare * GoogleAppMeasurement
Prepare * GoogleDataTransport
Prepare * GoogleMaps
Prepare * GooglePlaces
Prepare * GoogleUtilities
Prepare * GTMSessionFetcher
Prepare * Kingfisher
Prepare * leveldb-library
Prepare * libPhoneNumber-iOS
Prepare * MBProgressHUD
Prepare * nanopb
Prepare * OneSignal
Prepare * PromisesObjC
Prepare * SDWebImage
Prepare * SwiftyJSON
Prepare * TagListView
Prepare * UICKeyChainStore
Prepare Calculate checksums ✓
Prepare Resolving dependencies graph
Prepare Build targets (28):
Prepare * Alamofire
Prepare * DifferenceKit
Prepare * ECSlidingViewController
Prepare * Firebase
Prepare * FirebaseAnalytics
Prepare * FirebaseAuth
Prepare * FirebaseCore
Prepare * FirebaseCoreDiagnostics
Prepare * FirebaseDatabase
Prepare * FirebaseInstallations
Prepare * Google-Maps-iOS-Utils
Prepare * GoogleAppMeasurement
Prepare * GoogleDataTransport
Prepare * GoogleMaps
Prepare * GooglePlaces
Prepare * GoogleUtilities
Prepare * GTMSessionFetcher
Prepare * Kingfisher
Prepare * leveldb-library
Prepare * libPhoneNumber-iOS
Prepare * MBProgressHUD
Prepare * nanopb
Prepare * OneSignal
Prepare * PromisesObjC
Prepare * SDWebImage
Prepare * SwiftyJSON
Prepare * TagListView
Prepare * UICKeyChainStore
Prepare Add build target: RugbyPods
Prepare Save project ✓
[10s] Prepare ✓
Build Building sim-x86_64 ✓
Build Update checksums ✓
[5m 55s] Build ✓
Integration Update paths to built pods ✓
[1s] Integration ✓
Clean up Read project ✓
Clean up Remove sources from project
Clean up Remove frameworks
Clean up Remove products
Clean up Remove build target
Clean up Remove built pods
Clean up Remove schemes
Clean up Save project ✓
[3s] Clean up ✓
[6m 9s] Cached 28/28 pods.
[!] Project size ↓95%
[!] Indexing files count ↓99%
[!] Targets count ↓93%
[6m 9s] Let's roll 🏈

cat .rugby/history/28.02.2022T08.14.16/build.log 
[nanopb] Touching nanopb.framework
[libPhoneNumber-iOS] Touching libPhoneNumber_iOS.framework
[UICKeyChainStore] Touching UICKeyChainStore.framework
[leveldb-library] Touching leveldb.framework
[TagListView] Touching TagListView.framework
[SwiftyJSON] Touching SwiftyJSON.framework
[SDWebImage] Touching SDWebImage.framework
[PromisesObjC] Touching FBLPromises.framework
[MBProgressHUD] Touching MBProgressHUD.framework
[GoogleUtilities] Touching GoogleUtilities.framework
[GoogleDataTransport] Touching GoogleDataTransport.framework
[GTMSessionFetcher] Touching GTMSessionFetcher.framework
[Google-Maps-iOS-Utils] Touching GoogleMapsUtils.framework
[FirebaseCoreDiagnostics] Touching FirebaseCoreDiagnostics.framework
[Kingfisher] Touching Kingfisher.framework
[FirebaseCore] Touching FirebaseCore.framework
[ECSlidingViewController] Touching ECSlidingViewController.framework
[FirebaseInstallations] Touching FirebaseInstallations.framework
[FirebaseDatabase] Touching FirebaseDatabase.framework
[FirebaseAuth] Touching FirebaseAuth.framework
[DifferenceKit] Touching DifferenceKit.framework
[Alamofire] Touching Alamofire.framework
```

 - Podfile or Example project (if you can share these)
```
platform :ios, '12.0'

source 'https://cdn.cocoapods.org/'

target 'ProjectName' do
  # Comment the next line if you're not using Swift and don't want to use dynamic frameworks
  use_frameworks!
  
  # Pods for ProjectName
  pod 'GooglePlaces'
  pod 'GoogleMaps'
  pod 'Google-Maps-iOS-Utils', '~> 4.1.0'
  
  pod 'Firebase/Core', '~> 6'
  pod 'Firebase/Auth', '~> 6'
  pod 'Firebase/Database', '~> 6'

  pod 'Alamofire', '~> 4.7'

  pod 'OneSignal', '>= 3.0.0', '< 4.0'

  pod 'SwiftyJSON'
  pod 'DifferenceKit'
  pod 'UICKeyChainStore'
  pod 'Kingfisher', '~> 7.0'
  pod 'TagListView', '~> 1.0'
  pod 'MBProgressHUD', '~> 1.2.0'

  pod 'SDWebImage', '~> 4.0'
  pod 'ECSlidingViewController', '~> 2.0.3'
  pod 'libPhoneNumber-iOS', '~> 0.8'
  
  target 'OneSignalNotificationServiceExtension' do
    #only copy below line
    pod 'OneSignal', '>= 3.0.0', '< 4.0'
  end
  
end
```

## Additional context
I wanted to add that I can do archiving and uploading to TestFlight successfully with normal way without Rugby.

**Reply by devahmedshendy** [reply]
In case this should be turned to an issue, I made it ready for this.

**Reply by swiftyfinch** [reply]
Hello @devahmedshendy!
Did you try to use 🏈 Rugby like this: `rugby --config Release --sdk ios --bitcode`?
https://github.com/swiftyfinch/Rugby/blob/main/Docs/Cache.md#build-for-appstore-for-example

**Reply by devahmedshendy** [reply]
@swiftyfinch 
Oh I didn't search in this doc, only readme.

Should I repeat this each time I want to make new archive? Or just one time like I did with pods for development?

**Reply by swiftyfinch** [reply]
You need to run it each time before making a new archive. But Rugby caches this and if you don't have changes in dependencies that this call will be very fast. You can see cache information in `cache.yml` file.

**Reply by devahmedshendy** [reply]
@swiftyfinch 
Ok I see it worked, so another question.

When I ran this command first time it said:
```
⛔️ Error: Project already patched.
```

I understood that I should run it after clean pod installation, so should I use this command only before building the archive, or I can use it for development too? Sorry I am a bit confused.

**Reply by devahmedshendy** [reply]
From docs, I think I should work with `rugby` command only for development and once ready for archive I should reinstall pods again then execute that command.

I see from the doc also I can choose sdk for both sim and ios but still beginner to know what is the difference between Release and Debug configuration (This is my first time to release iOS app), I will learn about this and come back for any more questions.

**Reply by swiftyfinch** [reply]
> should I use this command only before building the archive, or I can use it for development too?

You can use that command for development, but without ability to use debugger and only for device.
That because in that command we passing `--config Release` and `--sdk ios`.
Also, running with `--bitcode` option will take more time. It's just not optimal for development.

> From docs, I think I should work with rugby command only for development and once ready for archive I should reinstall pods again then execute that command.

Rugby is just tool with different command and options, you can use it how it will be comfortable for you.
But I developed it for optimisation and I didn't think about releasing with Rugby. Now this options were added, but you can use it how you like.

> I see from the doc also I can choose sdk for both sim and ios but still beginner to know what is the difference between Release and Debug configuration (This is my first time to release iOS app), I will learn about this and come back for any more questions.

You are welcome, take you time.

**Reply by devahmedshendy** [reply]
@swiftyfinch 
So, is it possible that rugby would cache pods for both development and release? I mean, we already have this option with Xcode, we develop, then we archive, with same pods .. I thought to share this in case this is possible .. a little overload in caching (just one time) maybe an option for seamless development .. when we archive, we may archive after just a single bug fix.

**Reply by swiftyfinch** [reply]
When you switch to archive in Xcode, it also builds another version of your pods.
It's not the same as build for development. Rugby does the same.
So, Xcode also doesn't build all types of pods in one moment.

However, there is feature request #93, but I don't have enough free time to implement it.
And sorry, now is not the best time for me.

Also, you can use this solution https://github.com/swiftyfinch/Rugby/discussions/88#discussioncomment-1691293
But it's just a workaround.
